File: app/src/main/java/com/ivanovsky/passnotes/data/repository/file/FileSystemSyncProcessor.java
package com.ivanovsky.passnotes.data.repository.file;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.ivanovsky.passnotes.data.entity.ConflictResolutionStrategy;
import com.ivanovsky.passnotes.data.entity.FileDescriptor;
import com.ivanovsky.passnotes.data.entity.OperationResult;
import com.ivanovsky.passnotes.data.entity.SyncConflictInfo;
import com.ivanovsky.passnotes.data.entity.SyncProgressStatus;
import com.ivanovsky.passnotes.data.entity.SyncStatus;

public interface FileSystemSyncProcessor {

    @Nullable
    FileDescriptor getCachedFile(@NonNull String uid);

    @NonNull
    SyncProgressStatus getSyncProgressStatusForFile(@NonNull String uid);

    @NonNull
    SyncStatus getSyncStatusForFile(@NonNull String uid);

    @Nullable
    String getRevision(@NonNull String uid);

    @NonNull
    OperationResult<SyncConflictInfo> getSyncConflictForFile(@NonNull String uid);

    /** Returns updated FileDescriptor */
    @NonNull
    OperationResult<FileDescriptor> process(
            @NonNull FileDescriptor file,
            @NonNull SyncStrategy syncStrategy,
            @Nullable ConflictResolutionStrategy resolutionStrategy);
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/data/repository/file/FileSystemSyncProcessor.kt
package com.ivanovsky.passnotes.data.repository.file

import com.ivanovsky.passnotes.data.entity.FileDescriptor
import com.ivanovsky.passnotes.data.entity.MergeFiles
import com.ivanovsky.passnotes.data.entity.OperationResult
import com.ivanovsky.passnotes.data.entity.RequestedSyncResolution
import com.ivanovsky.passnotes.data.entity.SyncConflictInfo
import com.ivanovsky.passnotes.data.entity.SyncProgressStatus
import com.ivanovsky.passnotes.data.entity.SyncStatus

interface FileSystemSyncProcessor {

    fun getCachedFile(uid: String): FileDescriptor?

    fun getSyncProgressStatusForFile(uid: String): SyncProgressStatus

    fun getSyncStatusForFile(uid: String): SyncStatus

    fun getRevision(uid: String): String?

    fun getSyncConflictForFile(uid: String): OperationResult<SyncConflictInfo>

    fun getMergeFiles(uid: String): OperationResult<MergeFiles>

    fun process(
        file: FileDescriptor,
        requestedResolution: RequestedSyncResolution
    ): OperationResult<FileDescriptor>
}